src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,488 @@
+        {
+            await ErrorIdCannotBeBlankAsync(message);
+            return;
+        }
+        var messageId = message.Id!;
+
+        var dummyDisposer = new DummyDisposer();
+
+        try
+        {
+            if (overwrite)
+            {
+                Subscriptions[messageId] = dummyDisposer;
+            }
+            else
+            {
+                if (!Subscriptions.TryAdd(messageId, dummyDisposer))
+                {
+                    await ErrorIdAlreadyExistsAsync(message);
+                    return;
+                }
+            }
+
+            var result = await ExecuteRequestAsync(message);
+            CancellationToken.ThrowIfCancellationRequested();
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            if (result.Streams?.Count == 1)
+            {
+                // do not return a result, but set up a subscription
+                var stream = result.Streams!.Single().Value;
+                // note that this may immediately trigger some notifications
+                var disposer = stream.Subscribe(new Observer(this, messageId, _options.WebSockets.DisconnectAfterErrorEvent, _options.WebSockets.DisconnectAfterAnyError));
+                try
+                {
+                    if (Subscriptions.CompareExchange(messageId, dummyDisposer, disposer))
+                    {
+                        disposer = null;
+                    }
+                }
+                finally
+                {
+                    disposer?.Dispose();
+                }
+            }
+            else if (result.Executed)
+            {
+                await SendSingleResultAsync(message, result);
+            }
+            else
+            {
+                await SendErrorResultAsync(message, result);
+            }
+        }
+        catch (OperationCanceledException) when (CancellationToken.IsCancellationRequested)
+        {
+            throw;
+        }
+        catch (ExecutionError error)
+        {
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            await SendErrorResultAsync(message, error);
+        }
+        catch (Exception ex)
+        {
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
+            var error = await HandleErrorDuringSubscribeAsync(message, ex);
+            await SendErrorResultAsync(message, error);
+        }
+    }
+
+    /// <summary>
+    /// Creates an <see cref="ExecutionError"/> for an unknown <see cref="Exception"/>.
+    /// </summary>
+    protected virtual Task<ExecutionError> HandleErrorDuringSubscribeAsync(OperationMessage message, Exception ex)
+        => Task.FromResult<ExecutionError>(new UnhandledError("Unable to set up subscription for the requested field.", ex));
+
+    /// <summary>
+    /// Sends a single result to the client for a subscription or request, along with a notice
```

Comment:
> for a subscription **or** request

Or **only** for requests (query/mutation) ?

Reply:
Changed to "for a query or mutation request"